File: 2022/day10/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <vector>
using namespace std;

vector<string> split(string, string);

map<int, int> parseInput(ifstream& file) {
    map<int, int> obj;
    int X = 1;
    int cycle = 0;
    string line;
    while (getline(file, line)) {
        auto splitted = split(line, " ");
        if (splitted.size() == 1) { // noop
            cycle++;
            obj[cycle] = X;
        } else { // addx
            int amt = stoi(splitted[1]);
            cycle++;
            obj[cycle] = X;
            cycle++;
            obj[cycle] = X;
            X += amt;
        }
    }
    return obj;
}

int part1(map<int,int>& obj) {
    int s = 0;
    int nums[] = { 20, 60, 100, 140, 180, 220 };
    for (auto& x : nums) {
        s += x*obj[x];
    }
    return s;
}

void part2(map<int,int>& obj) {
    int pixelPos = 0;
    for (int r = 0; r < 6; r++) {
        for (int c = 0; c < 40; c++) {
            int spritePos = obj[pixelPos+1];
            if (spritePos-1 <= c && c <= spritePos+1) { // c is equivalent to pixelPos%40
                cout << "#" << " ";
            } else {
                cout << " " << " ";
            }
            pixelPos++;
        }
        cout << endl;
    }
}

int main(int argc, char** argv) {
    string infile = argc == 2 ? argv[1] : "../input.txt";
    ifstream file(infile);
    auto obj = parseInput(file);
    cout << "Part 1: " << part1(obj) << endl;
    cout << "Part 2:" << endl;
    part2(obj);
    file.close();
    return 0;
}

vector<string> split(string s, string delim) {
    size_t pos = 0;
    string token;
    vector<string> V;
    while ((pos = s.find(delim)) != string::npos) {
        token = s.substr(0, pos);
        V.push_back(token);
        s.erase(0, pos+delim.length());
    }
    V.push_back(s);
    return V;
}

File: 2022/day10/python/solution.py
import sys
infile = sys.argv[1] if len(sys.argv) > 1 else "../input.txt"

with open(infile) as f:
    X = 1
    cycle = 0
    obj = {}
    while line := f.readline().rstrip():
        op = line.split()[0]
        if op == "noop":
            cycle += 1
            obj[cycle] = X
        elif op == "addx":
            amt = int(line.split()[1])
            cycle += 1
            obj[cycle] = X
            cycle += 1
            obj[cycle] = X
            X += amt

def part1(obj):
    s = 0
    for i in [20, 60, 100, 140, 180, 220]:
        s += i*obj[i]
    return s

def part2(obj):
    img = [[' ' for _ in range(40) ] for _ in range(6)] # made ' ' the default value instead of '.' so that the final result is more readable
    pixelPos = 0
    for r in range(len(img)):
        for c in range(len(img[r])):
            spritePos = obj[pixelPos+1]
            if c in range(spritePos-1, spritePos+2): # c is the same thing as pixelPos%40
                img[r][c] = "#"
            pixelPos += 1
    return img

print("Part 1:", part1(obj))
print("Part 2:")
finalImg = part2(obj)
for r in finalImg:
    for c in r:
        print(c, end=" ")
    print()
